pkg/migration/migrations/public_schema_migration.go

```diff
@@ -0,0 +1,203 @@
+		if err != nil {
+			return err
+		}
+		if !found {
+			return errors.Newf("expected to find database with id %d", dbID)
+		}
+		mutableDesc := dbdesc.NewBuilder(desc.DatabaseDesc())
+		dbDesc := mutableDesc.BuildExistingMutable()
+
+		publicSchemaID, err := catalogkv.GenerateUniqueDescID(ctx, d.DB, d.Codec)
+		if err != nil {
+			return err
+		}
+		// Every database must be initialized with the public schema.
+		// Create the SchemaDescriptor.
+		// In postgres, the user "postgres" is the owner of the public schema in a
+		// newly created db. Postgres and Public have USAGE and CREATE privileges.
+		// In CockroachDB, root is our substitute for the postgres user.
+		publicSchemaPrivileges := descpb.NewBasePrivilegeDescriptor(security.AdminRoleName())
+		// By default, everyone has USAGE and CREATE on the public schema.
+		publicSchemaPrivileges.Grant(security.PublicRoleName(), privilege.List{privilege.CREATE, privilege.USAGE})
+		publicSchemaDesc := schemadesc.NewBuilder(&descpb.SchemaDescriptor{
+			ParentID:   dbID,
+			Name:       "public",
+			ID:         publicSchemaID,
+			Privileges: publicSchemaPrivileges,
+			Version:    1,
+		}).BuildCreatedMutableSchema()
+
+		descID := publicSchemaDesc.GetID()
+		idKey := catalogkeys.MakeSchemaNameKey(d.Codec, dbID, publicSchemaDesc.GetName())
+
+		key := catalogkeys.EncodeNameKey(d.Codec, catalogkeys.NewNameKeyComponents(dbID, keys.RootNamespaceID, tree.PublicSchema))
+
+		b := txn.NewBatch()
+
+		// Remove namespace entry for old public schema.
+		b.Del(key)
+		b.CPut(idKey, descID, nil)
+		if err := catalogkv.WriteNewDescToBatch(
+			ctx,
+			false,
+			d.Settings,
+			b,
+			d.Codec,
+			descID,
+			publicSchemaDesc,
+		); err != nil {
+			return err
+		}
+
+		dbDesc.DescriptorProto().GetDatabase().Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
+			tree.PublicSchema: {
+				ID: publicSchemaID,
+			},
+		}
+		dbDesc.MaybeIncrementVersion()
+		if err := catalogkv.WriteDescToBatch(
+			ctx, false, d.Settings, b, d.Codec, dbDesc.GetID(), dbDesc,
+		); err != nil {
+			return err
+		}
+
+		if err := migrateObjectsInDatabase(ctx, dbID, d, txn, b, publicSchemaID, descriptors); err != nil {
+			return err
+		}
+
+		return txn.Run(ctx, b)
+	})
+}
+
+func migrateObjectsInDatabase(
+	ctx context.Context,
+	dbID descpb.ID,
+	d migration.TenantDeps,
+	txn *kv.Txn,
+	batch *kv.Batch,
+	newPublicSchemaID descpb.ID,
+	descriptors *descs.Collection,
+) error {
+	descs, err := descriptors.GetAllDescriptorsInDatabase(ctx, txn, dbID)
+	if err != nil {
+		return err
+	}
+
+	var modifiedDescs []catalog.MutableDescriptor
+	for _, desc := range descs {
+		b := catalogkv.NewBuilder(desc.DescriptorProto())
+		switch b.DescriptorType() {
+		case catalog.Database, catalog.Schema:
+			panic(fmt.Sprintf("unexpected descriptor type %v", b.DescriptorType()))
+		case catalog.Table:
+			table := b.BuildExistingMutable()
```

Comment:
nit: do `switch mut := b.BuildExistingMutable().(type)` instead, and then you can access fields directly, like `mut.UnexposedParentSchemaID = newPublicSchemaID` in `case *tabledesc.Mutable` and `mut.ParentSchemaID = newPublicSchemaID` in `case *typedesc.Mutable`

Reply:
Good call, updated